enhancements/installer/aws-custom-edge-machineset-local-zones.md

```diff
@@ -0,0 +1,677 @@
+- Ensure the installer can create the MachineSet with `NoSchedule` taints on edge machine pools.
+- Ensure Local Zone subnets will not be used on `worker` compute pools or control planes.
+- Ensure the Wavelength zone will not be used in any compute pool
+
+Phase 2:
+
+- Admins can configure the Local availability zones before cluster creation.
+- Ensure the installer creates network resources based on the Local Zone defined on the
+  `edge` compute pool.
+
+### Non-Goals
+
+- Allow multiple compute pools for `edge`
+- Allow multiple instance types for the edge compute pool
+- Support private Local Zone subnets
+- Create any resource for Wavelength zones. Although Wavelength is also running on
+  the edge, it introduces a new resource called [Carrier Gateway](https://docs.aws.amazon.com/wavelength/latest/developerguide/carrier-gateways.html) to provide
+  connectivity between the compute nodes and the carrier network.
+
+Phase 1:
+
+- Create the network resources on the edge locations. That implementation will be
+  done in the second stage, when the installer will fully automate the creation
+  of resources on the availability zones specified on the `compute[.name=='edge'].platform.aws.zones`.
+
+## Proposal
+
+The phases proposed in this enhancement are:
+
+- Phase 1: Introduce the `edge` compute pool restricted to AWS Local Zones. Allow
+  users to install a cluster in existing networks with Local Zone subnets.
+- Phase 2: Create the network resources for AWS Local Zone locations.
+
+### Phase 1: Introduce the "edge" compute pool
+
+The first phase will introduce the new group of compute pool designed to
+be used exclusively for the user workloads on the zones not hosted in the
+default availability zones within the region, where the cluster is used to
+be deployed.
+
+Due to the limitation of cloud services in the edge locations, the dedicated
+pool also has limited compute and EBS type when compared to the default
+IPI (i.e. on AWS: m6i.xlarge).
+
+The edge machine pool will be available only when deploying on the AWS platform
+with Local Zones. In the future, it can be extended to the AWS Wavelength,
+and other cloud providers if needed.
+
+The following items must be satisfied in this phase:
+
+- the installer must allow one item of machine pool type `edge` (`compute[?name=="edge"]`)
+- the installer must extract Local Zone subnets from the subnet IDs
+  on `platform.aws.subnets`, creating one MachineSet manifest for each zone.
+- the `compute[?name=="edge"]` must create MachineSets manifests on Local Zone
+  subnets with:
+  - labels: `node-role.kubernetes.io/edge=''`, `location=local-zone`, and
+    `zone_group=zone_group_name`
+  - and taint `NoSchedule` matching the label `node-role.kubernetes.io/edge`
+- the MachineSet should be created when any Local Zone subnet is added on
+  `platform.aws.subnets`, even if compute pool was not defined on `compute[?name=="edge"]`
```

Comment:
> Why "even if"?

To keep compatibility with the existing flow when installing in existing VPC. The installer will look at each ZoneType of subnetIDs provided on `platform.aws.subnets`.

> Surely we are asking the user to specify the definition for the edge in compute[edge]?

Phase 1 covers only installation in existing VPC, thus the unique requirement is to provide the subnetID located in Local Zones. The user can provide the edge pool definition if they want to customize the pool, but may not be required. Thoughts?

>  How do we work out if platform.aws.subnets contains a local zone subnet?

The ZoneType attribute for each subnetID. [1]

> Is that something that is easy to look up?

Yes, currently there's a classification of ["public" or "private"](https://github.com/openshift/installer/blob/master/pkg/asset/installconfig/aws/subnet.go#L113-L121) subnets, the edge lookup will be added in the same flow.

> Feels like we are creating an implicit behaviour rather than being explicit about it

IIUC, the 'implicit behavior only' may not keep the compatibility with existing installer flow when installing in existing VPC, where it expects to create MachineSets on all AZs discovered by subnetID.

[1] https://github.com/openshift/enhancements/pull/1232/files#r979171225

Reply:
> To keep compatibility with the existing flow when installing in existing VPC. The installer will look at each ZoneType of subnetIDs provided on platform.aws.subnets.

That should be called out, it wasn't obvious to me that this provides backwards compatibility :)

Reply:
I created new sections under the [Proposal of Phase-1](https://github.com/openshift/enhancements/pull/1232/commits/fdbdba4c38f3b17c5159b4f5206a6a3671bc74ce#diff-f2590494c9138551ab5b3d5632d11170fe8742a128256f9e1d8ebc63baa19f41R178) to cover your items here. Thanks for sharing it. :)